### PR TITLE
HDDS-7039. EC: Handle the placement policy check in ECUnderReplicationHandler

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?

Currently ECUnderReplicationHandler did not handle the policy placement checks. This is the Jira to add that.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7039

## How was this patch tested?
Unit Tests

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java [hunk 1]
```diff
@@ -186,28 +209,30 @@ public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
...
-        // In this case we need to do one to one copy.
-        for (ContainerReplica replica : replicas) {
-          if (decomIndexes.contains(replica.getReplicaIndex())) {
-            if (!iterator.hasNext()) {
-              LOG.warn("Couldn't find enough targets. Available source"
-                  + " nodes: {}, the target nodes: {}, excluded nodes: {} and"
-                  + "  the decommission indexes: {}",
-                  replicas, selectedDatanodes, excludedNodes, decomIndexes);
-              break;
+        if (validatePlacement(nodes, selectedDatanodes)) {
```
Comment (umamaheswararao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:212]: Here nodes contains decom nodes as well? If we should we check excluding decom index nodes?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java [hunk 2]
```diff
@@ -73,6 +73,20 @@ public ECUnderReplicationHandler(
...
 
+  private boolean validatePlacement(List<DatanodeDetails> replicaNodes,
+                                    List<DatanodeDetails> selectedNodes) {
+    List<DatanodeDetails> nodes = new ArrayList<>(replicaNodes);
+    nodes.addAll(selectedNodes);
+    boolean placementStatus = containerPlacement
+            .validateContainerPlacement(nodes, nodes.size())
+            .isPolicySatisfied();
+    if (!placementStatus) {
+      LOG.warn("Selected Nodes does not satisfy placement policy: {} " +
```
Comment (umamaheswararao) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:84]: Typo: Nodes - nodes and please put a period before starting next sentence.

Comment (umamaheswararao) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:84]: Are you missed to pass the params here?

**Comment by umamaheswararao** [comment]
@swamirishi there are checkstyle warnings and update the patch?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:218]: Yeah, we should exclude them.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/protocol/DatanodeDetails.java [hunk 3]
```diff
@@ -234,6 +234,19 @@ public HddsProtos.NodeOperationalState getPersistedOpState() {
...
+  /**
+   * Checks if the OperationalState is Node is Decomissioned or Decomissioning.
+   * @return True if OperationalState is Decommissioned or Decomissioning.
+   */
+  public boolean isDecomissioned() {
+    return this.getPersistedOpState() ==
+            HddsProtos.NodeOperationalState.DECOMMISSIONED ||
+            this.getPersistedOpState() ==
+            HddsProtos.NodeOperationalState.DECOMMISSIONING;
+  }
```
Comment (umamaheswararao) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/protocol/DatanodeDetails.java:246]: Could you please avoid this multiple empty lines?

Comment (swamirishi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/protocol/DatanodeDetails.java:246]: done

**Review by umamaheswararao — APPROVED** [review]
LGTM

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java [hunk 4]
```diff
@@ -153,37 +166,48 @@ public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
         }
       }
       List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      Map<Integer, Pair<ContainerReplica, NodeStatus>> sources =
+              filterSources(replicas, deletionInFlight);
+      List<DatanodeDetails> nodes =
+              sources.values().stream().map(Pair::getLeft)
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:172]: I'm confused about this piece of code, but I may be missing something obvious.

There are two filters:

```
.filter(datanodeDetails ->
                              datanodeDetails.getPersistedOpState() ==
                              HddsProtos.NodeOperationalState.IN_SERVICE)
```

And

```
.filter(DatanodeDetails::isDecomissioned)
```
Are these not mutually exclusive? Ie, a node cannot be IN_SERVICE and isDecommissioned - so will this not filter everything out?

Inside filter sources, we already sort the IN_SERVICE above other states - I am wondering why we need to filter again here?

**Comment by sodonnel** [comment]
After this change, will mis-replication be detected even if there are no missing or decommission indexes? We only seem to call `validateContainerPlacement()` inside the if blocks for missing or decommission, so I don't think the handler will detect and schedule commands to fix mis-replication? I thought that this is what this Jira was to fix, but perhaps we need another Jira to do that check?

**Pull request merged**